src/filter.jl

```diff
@@ -85,3 +85,206 @@ function hamilton_filter(y::AbstractVector, h::Integer)
+"""
+mutable struct RegimeSwitchingModel{TD <: AbstractArray, TP <: AbstractMatrix, TPS <: AbstractMatrix}
+    g::Function
+    y::TD
+    parameter
+    P::TP
+    M::Int
+    prob_smoothed::TPS
+    logL::Float64
+    function RegimeSwitchingModel(g, y, parameter, P, M, prob_smoothed, logL)
+        if size(P, 1) != size(P, 2)
+            error("P must be square")
+        end
+        if size(P, 1) != M
+            error("the number of rows and columns of P must be equal to M")
+        end
+        return new{typeof(y), typeof(P), typeof(prob_smoothed)}(g, y, parameter, P, M, prob_smoothed, logL)
+    end
+end
+RegimeSwitchingModel(g::Function, y::AbstractArray, parameter, P::AbstractMatrix) =
+    RegimeSwitchingModel(g, y, parameter, P, size(P, 1), fill(NaN, size(y, 1), size(P, 1)), NaN)
+
+"""
+Apply the filter developed by Hamilton (1989, Econometrica) to a regime switching model.
+
+##### Arguments
+- `rsm::RegimeSwitchingModel`: `RegimeSwitchingModel`
+- `prob_update_pre::AbstractArray`: Probability distribution of state at period 0 conditional on the information up to
+                                   period 0.
+
+##### Returns
+- `p`: likelihood at each period
+- `p_s_update`: Probability of period `t` data conditional on the information up to period `t`.
+- `p_s_forecast`: Probability of period `t` data conditional on the information up to period `t-1`.
+"""
+function filter!(rsm::RegimeSwitchingModel;
+                 prob_update_pre::AbstractArray = stationary_distributions(MarkovChain(rsm.P))[1])
+    T = size(rsm.y, 1)
+    p_s_forecast = Matrix{Float64}(undef, T, rsm.M)
+    p_s_update = Matrix{Float64}(undef, T, rsm.M)
+    p, _, _ = filter!(rsm, p_s_update, p_s_forecast, prob_update_pre = prob_update_pre)
+    return p, p_s_update, p_s_forecast
+end
+
+"""
+Apply the filter developed by Hamilton (1989, Econometrica) to a regime switching model.
+
+##### Arguments
+- `rsm::RegimeSwitchingModel`: `RegimeSwitchingModel` specifying the model
+- `p_s_update::AbstractMatrix`: Probability of period `t` data conditional on the information up to period `t`.
+- `p_s_forecast::AbstractMatrix`: Probability of period `t` data conditional on the information up to period `t-1`.
+- `prob_update_pre::AbstractArray`: Probability distribution of state at period 0 conditional on the information up to
+                                   period 0.
+
+##### Returns
+- `p`: Likelihood at each period.
+- `p_s_update`: Probability of period `t` data conditional on the information up to period `t`.
+- `p_s_forecast`: Probability of period `t` data conditional on the information up to period `t-1`.
+"""
+function filter!(rsm::RegimeSwitchingModel, 
+                 p_s_update::AbstractMatrix, p_s_forecast::AbstractMatrix;
+                 prob_update_pre::AbstractArray = stationary_distributions(MarkovChain(rsm.P))[1])
+    g, y = rsm.g, rsm.y
+    T = size(y, 1)
+    p = Vector{Float64}(undef, T)
+    logL = 0
+    for t in 1:T

```

Comment:
I like how the code is structured that allows you to write such a simple (and beautiful!) for-loop to do the core of the computation.